nixio/validator.py

```diff
@@ -0,0 +1,498 @@
+                            "atomic SI unit "
+                            "(Note: composite units are not supported)")
+    NoSamplingInterval = ("sampling interval for dimension {} "
+                          "is not set")
+    InvalidSamplingInterval = ("sampling interval for dimension {} "
+                               "is not valid (interval > 0)")
+    NoData = "data is not set"
+    NoLinkType = "link_type is not set"
+
+
+class ValidationWarning(object):
+    NoVersion = "version is not set"
+    NoFormat = "format is not set"
+    InvalidUnit = "unit is not SI or composite of SI units"
+    NoExpansionOrigin = ("polynomial coefficients for calibration are set, "
+                         "but expansion origin is missing")
+    NoPolynomialCoefficients = ("expansion origin for calibration is set, "
+                                "but polynomial coefficients are missing")
+    OffsetNoUnit = ("offset for dimension {} is set, "
+                    "but no valid unit is set")
+    NoUnit = "unit is not set"
+
+
+def check_file(nixfile):
+    """
+    Validate a NIX file and all contained objects and return all errors and
+    warnings for each individual object.
+
+    :returns: A nested dictionary of errors and warnings. Each subdictionary
+    is indexed by the object with values describing the error or warning.
+    :rtype: Dictionary
+    """
+    results = {"errors": dict(), "warnings": dict()}
+    if not nixfile.created_at:
+        results["errors"][nixfile] = [ValidationError.NoDate]
+
+    file_warnings = list()
+    if not nixfile.version:
+        file_warnings.append(ValidationWarning.NoVersion)
+    if not nixfile.format:
+        file_warnings.append(ValidationWarning.NoFormat)
+    if file_warnings:
+        results["warnings"][nixfile] = file_warnings
+
+    def update_results(obj, errors, warnings):
+        if errors:
+            results["errors"][obj] = errors
+        if warnings:
+            results["warnings"][obj] = warnings
+
+    # Blocks
+    for block in nixfile.blocks:
+        blk_errors, blk_warnings = check_block(block)
+        update_results(block, blk_errors, blk_warnings)
+
+        # Groups
+        for group in block.groups:
+            grp_errors, grp_warnings = check_group(group)
+            update_results(group, grp_errors, grp_warnings)
+
+        # DataArrays
+        for da in block.data_arrays:
+            da_errors, da_warnings = check_data_array(da)
+            update_results(da, da_errors, da_warnings)
+
+        # Tags
+        for tag in block.tags:
+            tag_errors, tag_warnings = check_tag(tag)
+            update_results(tag, tag_errors, tag_warnings)
+
+        # MultiTags
+        for mtag in block.multi_tags:
+            mtag_errors, mtag_warnings = check_multi_tag(mtag)
+            update_results(mtag, mtag_errors, mtag_warnings)
+
+        # Sources
+        def traverse_sources(sources):
```

Comment:
is this on the right indentation level?

Reply:
I guess it could be outside the loop, but inside the outer method.

Reply:
agreed, inside of ``check_file`` but outside of loop sounds good to me. Better than having method definition inside the loop, even though it will be usually only a few blocks in a file